packages/flutter/lib/src/widgets/fade_in_image.dart

```diff
@@ -416,9 +412,24 @@ class _FadeInImageState extends State<FadeInImage> {
       fit: widget.fit,
       frameBuilder: (BuildContext context, Widget child, int? frame, bool wasSynchronouslyLoaded) {
         if (wasSynchronouslyLoaded) {
-          _resetAnimations();
+          _oldImage = child;
           return child;
         }
+        if (_imageAnimation.isCompleted) {
```

Comment:
Is all this necessary? I naively expected that we would just need an instance variable (e.g. `targetLoaded = false`) and set that to true the first time `frame != null`. Then on line 443, we just pass `targetLoaded` to `isTargetLoaded`.

Maybe I am missing something, though?

Reply:
1. If you make the above-suggested change, run the sample code (attached in the issue), and you change the image (by clicking) a few times, after 4 changes (clicks) the view goes blank, and nothing will happen after any further clicks. This is because, after 3 clicks, the image changes to a loaded image, setting `wasSynchronouslyLoaded` to true, disposing `_AnimatedFadeOutFadeIn` and the animations. Now when the 4th click occurs, `_AnimatedFadeOutFadeIn` is rebuilt, resetting the opacity animations. This can be fixed by removing the `wasSynchronouslyLoaded` if statement, so that `_AnimatedFadeOutFadeIn` is never disposed.
2. IMO the `_AnimatedFadeOutFadeIn` widget should be added to the tree only when the actual fade-out-fade-in is required, and removed when not. The above change (along with point 1) will cause it to exist on the tree forever.

Reply:
It would be more performant by keeping the subtree the same, otherwise, they may be built from scratch

Reply:
I can't reproduce the issue you mention after 4 clicks. This is the code i use

```dart
frameBuilder: (BuildContext context, Widget child, int? frame, bool wasSynchronouslyLoaded) {
        if (wasSynchronouslyLoaded) {
          _oldImage = child;
          return child;
        }
        if (frame != null) {
          // This is the first image.
          _oldImage = child;
        }
        return _AnimatedFadeOutFadeIn(
          target: child,
          targetProxyAnimation: _imageAnimation,
          placeholder: _image(
            image: widget.placeholder,
            errorBuilder: widget.placeholderErrorBuilder,
            opacity: _placeholderAnimation,
            fit: widget.placeholderFit ?? widget.fit,
          ),
          placeholderProxyAnimation: _placeholderAnimation,
          isTargetLoaded: _oldImage != null,
          fadeInDuration: widget.fadeInDuration,
          fadeOutDuration: widget.fadeOutDuration,
          fadeInCurve: widget.fadeInCurve,
          fadeOutCurve: widget.fadeOutCurve,
        );
      },
```

The issue with the current code is that after the image is faded in, it will cause a   an unnecessary rebuild to dispose _AnimatedFadeOutFadeIn and rebuild the image widget which is not ideal

Reply:
you'll be able to reproduce the issue with below:

> we would just need an instance variable (e.g. targetLoaded = false) and set that to true the first time frame != null. Then on line 443, we just pass targetLoaded to isTargetLoaded.

<details>
<summary>code</summary>

```dart
  // Widget? _oldImage;
  bool targetLoaded = false;
```
...
```dart
    Widget result = _image(
      image: widget.image,
      errorBuilder: widget.imageErrorBuilder,
      opacity: _imageAnimation,
      fit: widget.fit,
      frameBuilder: (BuildContext context, Widget child, int? frame, bool wasSynchronouslyLoaded) {
        if (wasSynchronouslyLoaded) {
          return child;
        }
        if (frame != null) {
          targetLoaded = true;
        }
        return _AnimatedFadeOutFadeIn(
          target: child,
          targetProxyAnimation: _imageAnimation,
          placeholder: _image(
            image: widget.placeholder,
            errorBuilder: widget.placeholderErrorBuilder,
            opacity: _placeholderAnimation,
            fit: widget.placeholderFit ?? widget.fit,
          ),
          placeholderProxyAnimation: _placeholderAnimation,
          isTargetLoaded: targetLoaded, // _oldImage != null
          fadeInDuration: widget.fadeInDuration,
          fadeOutDuration: widget.fadeOutDuration,
          fadeInCurve: widget.fadeInCurve,
          fadeOutCurve: widget.fadeOutCurve,
        );
      },
    );
```
</details>

Reply:
> disposing _AnimatedFadeOutFadeIn and the animations. Now when the 4th click occurs, _AnimatedFadeOutFadeIn is rebuilt, resetting the opacity animations.

Wheere does the animations is disposed? It looks like we do need to reset the animation when the `wasSynchronouslyLoaded` is true to prepare for future rebuilds?

Reply:
> Where does the animations is disposed?

my bad, animations are not disposed, just the widget.

> It looks like we do need to reset the animation when the wasSynchronouslyLoaded is true to prepare for future rebuilds?

As per gapless playback, there won't be any future animations. The old image will just be replaced by the new one after it is loaded.